### ISSUE TITLE
Position information incorrect for a macro call that wrongly uses a semicolon

### ISSUE THREAD

**Issue by dmalyuta** [julia-vscode/StaticLint.jl#279]
From my understanding, macros can't use keyword arguments so a macro call can't have a semicolon in it. However, consider the following code where I (accidentally) used a semicolon:

![image](https://user-images.githubusercontent.com/12475811/116296239-d5221380-a74e-11eb-9dff-9bf14dc4a1c7.png)

There are two calls to `@variable`. Both use an undefined variable `bar`, which should produce an error. The first call is "correct" in that it uses a comma, while the second call is "incorrect" because it uses a semicolon. In both cases, we would expect `bar` to be highlighted as a missing reference error. Possibly we would also want to highlight the semicolon as inappropriate in the second call.

However, what I am seeing instead (on the current master branch latest commit) is that the position gets jumbled after the semicolon in the second `@variable` call. Also, somehow the missing reference error appears twice for the second call. Even worse, the errors in the subsequent code also are affected, and all appear in the wrong positions.

The code that I used to find the errors in the above file is below. Running it will print the errors and their start/end character positions.

```
using StaticLint
using StaticLint: quoted, unquoted, should_mark_missing_getfield_ref, hasref,
    parentof, is_in_fexpr, errorof, haserror

using CSTParser
using CSTParser: EXPR, headof, isidentifier, isnonstdid, valof, iscall

using SymbolServer
using Printf
using Sockets

# ..:: Variables and data structures ::..

mutable struct ErrorSpan
    beg_char::Integer
    end_char::Integer
end

SL = StaticLint
SS = SymbolServer
CSTP = SL.CSTParser
T_Error = Tuple{ErrorSpan,EXPR}

error = "error"
warn = "warn"
info = "info"

const LintCodeToErrorLevel = Dict{SL.LintCodes,String}(
    SL.IncorrectCallArgs => error,
    SL.IncorrectIterSpec => error,
    SL.NothingEquality => error,
    SL.NothingNotEq => error,
    SL.ConstIfCondition => info,
    SL.EqInIfConditional => error,
    SL.PointlessOR => info,
    SL.PointlessAND => info,
    SL.UnusedBinding => info,
    SL.InvalidTypeDeclaration => error,
    SL.UnusedTypeParameter => info,
    SL.IncludeLoop => warn,
    SL.MissingFile => error,
    SL.InvalidModuleName => error,
    SL.TypePiracy => warn,
    SL.UnusedFunctionArgument => warn,
    SL.CannotDeclareConst => error,
    SL.InvalidRedefofConst => error,
    SL.NotEqDef => warn,
    SL.KwDefaultMismatch => error,
    SL.InappropriateUseOfLiteral => warn,
    SL.ShouldBeInALoop => warn,
    SL.TypeDeclOnGlobalVariable => error,
    SL.UnsupportedConstLocalVariable => warn,
    SL.UnassignedKeywordArgument => error,
    SL.CannotDefineFuncAlreadyHasValue => error,
    SL.DuplicateFuncArgName => error)

# ..:: Functions ::..

#= Print out the error.

Args:
    p: the file path.
    pos: the (start,end) character positions.
    desc: the error description
    lvl: the error level (error? warning? info?)

Returns:
    msg: the error message in the agreed format. =#
function format_error(p::String, pos::ErrorSpan, desc::String, lvl::String)::String
    msg = @sprintf "%s:%d:%d: %s: %s\n" p pos.beg_char pos.end_char lvl desc
    return msg
end

#= Collect linter output about the file.

This is basically copied from the StaticLint.collect_hints function, however I
updated it slightly to output the (start, end) character positions of the
error, and incorporated the feedback from
https://github.com/julia-vscode/CSTParser.jl/issues/238 on how to make the
character positions match correctly.

Note: the second output is used internally by the function. You don't need to
look at it when using the function to collect the array of errors (the first
output). =#
function lint_collect_hints(x::EXPR,
                            server,
                            missingrefs=:all,
                            isquoted=false,
                            errs=T_Error[],
                            pos=1)::Tuple{Array{T_Error,1},Integer}
    if isquoted && unquoted(x)
        isquoted = false
    elseif quoted(x)
        isquoted = true
    end

    if headof(x) === :errortoken
        # collect parse errors
        push!(errs, (ErrorSpan(pos, pos+x.span), x))
    elseif !isquoted
        if missingrefs != :none &&
            isidentifier(x) &&
            !hasref(x) &&
            !(valof(x) == "var" &&
              parentof(x) isa EXPR &&
              isnonstdid(parentof(x))) &&
              !((valof(x) == "stdcall" ||
                 valof(x) == "cdecl" ||
                 valof(x) == "cdecl" ||
                 valof(x) == "fastcall" ||
                 valof(x) == "thiscall" ||
                 valof(x) == "llvmcall") &&
                is_in_fexpr(x, x -> (iscall(x) &&
                                     isidentifier(x.args[1]) &&
                                     valof(x.args[1]) == "ccall")))
            push!(errs, (ErrorSpan(pos, pos+x.span), x))
        elseif haserror(x) && errorof(x) isa SL.LintCodes
            # collect lint hints
            push!(errs, (ErrorSpan(pos, pos+x.span), x))
        end
    elseif (isquoted &&
            missingrefs == :all &&
            should_mark_missing_getfield_ref(x, server))
        push!(errs, (ErrorSpan(pos, pos+x.span), x))
    end

    for a in x
        if a.args === nothing
            lint_collect_hints(a, server, missingrefs, isquoted, errs, pos)
            pos += a.fullspan
        else
            _,pos = lint_collect_hints(a, server, missingrefs, isquoted, errs, pos)
        end
    end

    return errs, pos
end

#= Convert byte to character count.

Convert error start and end character locations from a byte measurement to a
strict character count measurement.

Args:
    src: the source code (as a string) which this error is for.
    offset: the error location in bytes. Modified in place. =#
function convert_pos_byte_to_char!(src::String, offset::ErrorSpan)
    strlen_byte = lastindex(src)
    offset.beg_char = length(src, 1, min(offset.beg_char, strlen_byte))
    offset.end_char = length(src, 1, min(offset.end_char, strlen_byte))
end

#= Convert error to an error level.

Args:
    err: the error object.

Returns:
    lvl: the error level. =#
function get_error_level(x::EXPR)::String
    error_code = SL.errorof(x)
    lvl = error # Default error level
    if haskey(LintCodeToErrorLevel, error_code)
        lvl = LintCodeToErrorLevel[error_code]
    end
    return lvl
end

#= Static lint a file.

Args:
    rootfile: the file to be linted.
    server: the StaticLint server.
    conn: the TCP connection to send errors back over. =#
function lint_file(rootfile::String,
                   server::SL.FileServer)::Nothing
    empty!(server.files)

    # Load and parse the file
    f = SL.loadfile(server, rootfile)

    # SL's main run- finding variables, scopes, etc.
    SL.semantic_pass(f)

    # Run lint checks.
    hints = Dict()
    slopts = SL.LintOptions(:)
    for (path, file) in server.files
        SL.check_all(file.cst, slopts, server)
        hints[path],_ = lint_collect_hints(file.cst, server)
    end

    # Send errors back to client
    files_src = Dict(f.path=>f.source)
    for (p, hs) in hints
        for (offset, x) in hs
            if !haskey(files_src, p)
                files_src[p] = read(p, String)
            end

            if (SL.haserror(x) &&
                SL.errorof(x) isa SL.LintCodes)
                description = SL.LintCodeDescriptions[SL.errorof(x)]
                level = get_error_level(x)
                convert_pos_byte_to_char!(files_src[p], offset)
                println(format_error(p, offset, description, level))
            else
                # Use of undeclared variable
                description = string("Missing reference for ", CSTP.valof(x))
                level = error
                convert_pos_byte_to_char!(files_src[p], offset)
                println(format_error(p, offset, description, level))
            end
        end
    end

    return nothing
end

# ..:: Start the server loop ::..

# Julia configuration
depot = first(SS.Pkg.depots())
cache = joinpath(dirname(pathof(SS)), "..", "store")
env = dirname(SS.Pkg.Types.Context().env.project_file)

# Setup server
server = SL.FileServer()
ssi = SymbolServerInstance(depot, cache)
_, server.symbolserver = SS.getstore(ssi, env)
server.symbol_extends  = SS.collect_extended_methods(server.symbolserver)

lint_file("test.jl", server)
```

```
using JuMP, ECOS

mdl = Model(ECOS.Optimizer)

@variable(mdl, base_name=bar)
@variable(mdl; base_name=bar)

morecode

somemorecode
```

The output is (the position errors begin with the second line):
```
test.jl:73:76: error: Missing reference for bar

test.jl:98:101: error: Missing reference for bar

test.jl:111:114: error: Missing reference for bar

test.jl:117:125: error: Missing reference for morecode

test.jl:127:131: error: Missing reference for somemorecode
```
